### ISSUE TITLE
Set default callback as last argument from caller function.

### ISSUE THREAD

**Issue by vjpr** [ashtuchkin/errTo#3]
Another improvement might be to allow a default callback being the last argument the caller function using `arguments.callee.caller`.

```
index: (req, res, next) ->
  await foo errTo defer bar
  res.send bar
```

Default callback would be used if only one argument is passed to `errTo`, in this case it is next.

Will need to check if this is compatible with IcedCoffee too.

**Comment by ashtuchkin** [comment]
Sorry, but I believe this will not work. In IcedCoffee (and in the usual callback style) the place where you call errTo is usually several functions deep from the original callback (you can see it in the raw js code). And also both callee and caller properties are deprecated in ES5.

**Comment by vjpr** [comment]
I'm sure there must be a way to get access to the original scope though, by accessing some of IcedCoffee's helper vars.

**Comment by vjpr** [comment]
I had a poke around and I found a hacky way to do it by walking up the call stack starting from `errTo.caller`, and using a regex to check if the last argument of the function is called `$done`.

```
_ = require 'underscore'

debug = false

foo = (done) ->
  done('foo err')
  # done()

bar = (done) ->
  # done('bar err')
  done()

errTo = (errorHandler, successHandler) ->

  if debug 
    console.log '\n\n---------------------'
    console.log 'errTo CALLED!'

  unless successHandler?

    # Find first non-Iced callback.
    fn = errTo.caller
    until /^function \(.*\$done\)/.test fn.toString()
      fn = fn.caller
      break unless fn?
      if debug
        console.log '---------------------'
        console.log fn.toString()

    unless fn?
      msg = 'No error handler named `$done` found'
      console.error msg
      throw new Error msg

    if debug
      console.log '---------------------' 
      console.log 'Found:', fn.toString()

    errCb = _.last fn.arguments

    # Update handlers.
    successHandler = errorHandler
    errorHandler = errCb

  (err, args...) ->
    return errorHandler err if err
    successHandler args...

index = (req, res, $done) ->
  await foo errTo defer()
  foo ->
    await bar errTo defer()
    $done()

index 'a', 'b', (err, val) ->
  return console.error('error:', err) if err
  console.log 'success:', val
```

**Comment by vjpr** [comment]
Just noticed your `noErr = errTo.bind null, done` trick. I like it.

My approach above seems way too hacky for saving one line of code.

**Comment by ashtuchkin** [comment]
Wow, that's an interesting idea. It even have an idea how to make it work for the asynchronous case where there are separate stacks. I'll try to implement it.

The thing about one line of code is that this line will have to be in every function, which is a no-no :)

**Comment by vjpr** [comment]
Cool! Look forward to it. The code I sent should work if you copy and paste
and npm install underscore.

**Comment by ashtuchkin** [comment]
Vaughan, I tried to implement it but unfortunately it takes longer than I anticipated. Today I've reviewed my todo list and it seems I will not be able to do it anytime soon, sorry.

If you would like to go further, I would suggest making it a separate module (`errUp`?).
Unfortunately its not likely that I accept a pull request because there are just too many cases I see that need to be checked and even if there's an extensive test coverage, it'll be hard for me to check the test cases and whether they are complete.

If you still would like to proceed, my another idea was when going up the stack, not only check for the parameter name, but also that the last parameter might be a result of calling another errTo or errUp, so that the following would work:

``` coffeescript
fn = ($done) ->
  doSomething1 errUp ->  # This one will catch up the parameter name
    doSomething2 errUp ->  # This one will catch that we were called from the previous errUp and get error handler from it.
      $done(null)
```
